File: apps/web/test-taxonomy.mjs
import assert from 'node:assert/strict';
import { DatabaseSync } from 'node:sqlite';

console.log('🧪 RUNNING CATEGORY TAXONOMY & NORMALIZATION TEST SUITE...\n');

const db = new DatabaseSync('../../database/tech_sentinel.db');

const CANONICAL_SYNONYMS = {
  'ai': 'ai',
  'artificial intelligence': 'ai',
  'artificial-intelligence': 'ai',
  'artificial_intelligence': 'ai',
  'machine learning': 'ai',
  'machine-learning': 'ai',
  'machine_learning': 'ai',
  'ml': 'ai',
  'genai': 'ai',
  'generative ai': 'ai',
  'deep learning': 'ai',
  'deep-learning': 'ai',
  'deep_learning': 'ai',
  'llm': 'ai',
  'llms': 'ai',
  'cloud': 'cloud',
  'cloud computing': 'cloud',
  'cloud-computing': 'cloud',
  'cloud_computing': 'cloud',
  'cloud infrastructure': 'cloud',
  'cloud-infrastructure': 'cloud',
  'cloud_infrastructure': 'cloud',
  'devops': 'cloud',
  'serverless': 'cloud',
  'infrastructure': 'cloud',
  'cloud & infrastructure': 'cloud',
  'cloud-native': 'cloud',
  'development': 'development',
  'software development': 'development',
  'software-development': 'development',
  'software_development': 'development',
  'software dev': 'development',
  'software-dev': 'development',
  'software_dev': 'development',
  'programming': 'development',
  'software engineering': 'development',
  'coding': 'development',
  'web development': 'development',
  'webdev': 'development',
  'frontend': 'development',
  'backend': 'development',
  'fullstack': 'development',
  'developer tools': 'development',
  'devtools': 'development',
  'open_source': 'open_source',
  'open source': 'open_source',
  'open-source': 'open_source',
  'opensource': 'open_source',
  'oss': 'open_source',
  'foss': 'open_source',
  'free and open source': 'open_source',
  'cybersecurity': 'cybersecurity',
  'cyber security': 'cybersecurity',
  'cyber-security': 'cybersecurity',
  'cyber_security': 'cybersecurity',
  'security': 'cybersecurity',
  'infosec': 'cybersecurity',
  'appsec': 'cybersecurity',
  'startups': 'startups',
  'startup': 'startups',
  'start-ups': 'startups',
  'start_ups': 'startups',
  'start-up': 'startups',
  'start_up': 'startups',
  'venture': 'startups',
  'entrepreneurship': 'startups',
  'education': 'education',
  'learning': 'education',
  'student': 'education',
  'students': 'education',
  'training': 'education',
  'tutorial': 'education',
  'tutorials': 'education',
};

function normalizeCategory(category) {
  if (!category || typeof category !== 'string') return 'development';
  const cleaned = category.trim().toLowerCase();
  const cleanedSpaces = cleaned.replace(/[\s_-]+/g, ' ').trim();
  const cleanedUnderscores = cleaned.replace(/[\s_-]+/g, '_').trim();

  if (CANONICAL_SYNONYMS[cleaned]) return CANONICAL_SYNONYMS[cleaned];
  if (CANONICAL_SYNONYMS[cleanedSpaces]) return CANONICAL_SYNONYMS[cleanedSpaces];
  if (CANONICAL_SYNONYMS[cleanedUnderscores]) return CANONICAL_SYNONYMS[cleanedUnderscores];

  const slug = cleaned.replace(/[^\w\s-]/g, '').trim().replace(/[\s-]+/g, '_');
  return slug || 'development';
}

// -------------------------------------------------------------
// Test 1: Category synonyms normalize to canonical slugs
// -------------------------------------------------------------
{
  assert.equal(normalizeCategory('AI'), 'ai');
  assert.equal(normalizeCategory('Artificial Intelligence'), 'ai');
  assert.equal(normalizeCategory('artificial-intelligence'), 'ai');
  assert.equal(normalizeCategory('machine learning'), 'ai');
  assert.equal(normalizeCategory('ML'), 'ai');

  assert.equal(normalizeCategory('Cloud Computing'), 'cloud');
  assert.equal(normalizeCategory('cloud-computing'), 'cloud');
  assert.equal(normalizeCategory('Cloud Infrastructure'), 'cloud');
  assert.equal(normalizeCategory('DevOps'), 'cloud');

  assert.equal(normalizeCategory('Software Development'), 'development');
  assert.equal(normalizeCategory('software-dev'), 'development');
  assert.equal(normalizeCategory('Programming'), 'development');
  assert.equal(normalizeCategory('Developer Tools'), 'development');

  assert.equal(normalizeCategory('Open Source'), 'open_source');
  assert.equal(normalizeCategory('OSS'), 'open_source');
  assert.equal(normalizeCategory('FOSS'), 'open_source');

  assert.equal(normalizeCategory('Cybersecurity'), 'cybersecurity');
  assert.equal(normalizeCategory('Cyber Security'), 'cybersecurity');
  assert.equal(normalizeCategory('Security'), 'cybersecurity');

  assert.equal(normalizeCategory('Startups'), 'startups');
  assert.equal(normalizeCategory('start-up'), 'startups');

  console.log('✅ Test 1 Passed: Equivalent category variations successfully normalize to canonical taxonomy slugs.');
}

// -------------------------------------------------------------
// Test 2: Unrelated categories are never improperly merged
// -------------------------------------------------------------
{
  const distinct = ['ai', 'cloud', 'development', 'open_source', 'cybersecurity', 'startups', 'education'];
  const normalized = distinct.map(c => normalizeCategory(c));
  const uniqueNormalized = new Set(normalized);
  
  assert.equal(uniqueNormalized.size, distinct.length, 'All distinct taxonomy categories must remain distinct');
  console.log(`✅ Test 2 Passed: ${distinct.length} distinct technology domains remain strictly separate without over-normalization.`);
}

// -------------------------------------------------------------
// Test 3: Active SQLite database categories are 100% canonical
// -------------------------------------------------------------
{
  const rows = db.prepare(`
    SELECT DISTINCT category 
    FROM news 
    WHERE category IS NOT NULL AND TRIM(category) != ''
  `).all();

  for (const r of rows) {
    const norm = normalizeCategory(r.category);
    assert.equal(r.category, norm, `Category in DB '${r.category}' must already be canonical (expected '${norm}')`);
  }
  console.log(`✅ Test 3 Passed: Verified all ${rows.length} distinct categories in SQLite news records match canonical taxonomy: [${rows.map(r => r.category).join(', ')}].`);
}

// -------------------------------------------------------------
// Test 4: Query filtering works transparently with raw synonym inputs
// -------------------------------------------------------------
{
  // Searching with 'Artificial Intelligence' should query canonical 'ai' records
  const normInput = normalizeCategory('Artificial Intelligence');
  const aiRecords = db.prepare('SELECT * FROM news WHERE category = ?').all(normInput);
  assert(aiRecords.length > 0, "Querying with synonym 'Artificial Intelligence' must retrieve matching AI records");

  console.log(`✅ Test 4 Passed: Query filtering with synonym 'Artificial Intelligence' seamlessly retrieved ${aiRecords.length} canonical 'ai' records.`);
}

console.log('\n🎉 ALL CATEGORY TAXONOMY & NORMALIZATION TESTS PASSED PERFECTLY!\n');

File: tests/test_taxonomy.py
import unittest
from agent.utils.taxonomy import normalize_category
from agent.processors.classifier import ContentClassifier
from agent.models import RawItem

class TestCategoryTaxonomy(unittest.TestCase):
    def test_canonical_equivalents_normalized(self):
        # AI variations
        self.assertEqual(normalize_category("AI"), "ai")
        self.assertEqual(normalize_category("Artificial Intelligence"), "ai")
        self.assertEqual(normalize_category("artificial-intelligence"), "ai")
        self.assertEqual(normalize_category("machine learning"), "ai")
        self.assertEqual(normalize_category("ML"), "ai")
        self.assertEqual(normalize_category("deep_learning"), "ai")

        # Cloud variations
        self.assertEqual(normalize_category("Cloud"), "cloud")
        self.assertEqual(normalize_category("Cloud Computing"), "cloud")
        self.assertEqual(normalize_category("cloud-computing"), "cloud")
        self.assertEqual(normalize_category("Cloud Infrastructure"), "cloud")
        self.assertEqual(normalize_category("DevOps"), "cloud")
        self.assertEqual(normalize_category("Serverless"), "cloud")

        # Development variations
        self.assertEqual(normalize_category("Development"), "development")
        self.assertEqual(normalize_category("Software Development"), "development")
        self.assertEqual(normalize_category("software-dev"), "development")
        self.assertEqual(normalize_category("Programming"), "development")
        self.assertEqual(normalize_category("coding"), "development")
        self.assertEqual(normalize_category("Developer Tools"), "development")

        # Open Source variations
        self.assertEqual(normalize_category("Open Source"), "open_source")
        self.assertEqual(normalize_category("open-source"), "open_source")
        self.assertEqual(normalize_category("OSS"), "open_source")
        self.assertEqual(normalize_category("FOSS"), "open_source")

        # Cybersecurity variations
        self.assertEqual(normalize_category("Cybersecurity"), "cybersecurity")
        self.assertEqual(normalize_category("Cyber Security"), "cybersecurity")
        self.assertEqual(normalize_category("cyber-security"), "cybersecurity")
        self.assertEqual(normalize_category("Security"), "cybersecurity")
        self.assertEqual(normalize_category("infosec"), "cybersecurity")

        # Startups variations
        self.assertEqual(normalize_category("Startup"), "startups")
        self.assertEqual(normalize_category("Startups"), "startups")
        self.assertEqual(normalize_category("start-up"), "startups")
        self.assertEqual(normalize_category("venture"), "startups")

    def test_distinct_domains_not_merged(self):
        # AI and Cloud must remain distinct
        self.assertNotEqual(normalize_category("ai"), normalize_category("cloud"))
        self.assertNotEqual(normalize_category("cybersecurity"), normalize_category("development"))
        self.assertNotEqual(normalize_category("open_source"), normalize_category("startups"))

    def test_new_future_categories_preserved(self):
        # Novel distinct categories must form clean slugs
        self.assertEqual(normalize_category("Robotics"), "robotics")
        self.assertEqual(normalize_category("Quantum Computing"), "quantum_computing")
        self.assertEqual(normalize_category("Hardware & Embedded"), "hardware_embedded")

    def test_content_classifier_assigns_canonical_category(self):
        classifier = ContentClassifier()
        item = RawItem(
            title="Kubernetes 1.32 Cloud Architecture Guide",
            description="Scaling serverless cluster nodes in AWS and GCP",
            url="https://example.com/k8s-cloud",
            source_id="src_test",
            source_name="Cloud Blog"
        )
        cat, tags = classifier.classify(item)
        self.assertEqual(cat, "cloud")

if __name__ == "__main__":
    unittest.main()
